### PR TITLE
optimize the route connectivity

### PR DESCRIPTION
**Pull request by yaocw2020**
- Add connectivity initialization
- Add connectivity status "DHCP failed" and "ping failed", delete
  "unknown"
- Add pinger timeout and allow 20% package loss

Related issue: https://github.com/harvester/harvester/issues/1476

### CODE REVIEW

File: pkg/controller/manager/nad/controller.go [hunk 1]
```diff
@@ -97,22 +99,23 @@ func (h Handler) OnChange(key string, nad *cniv1.NetworkAttachmentDefinition) (*
...
 	klog.Infof("netconf: %+v", networkConf)
 
 	if networkConf.CIDR != "" && networkConf.Gateway != "" {
-		// set connectivity as the initial status unknown
+		// initialize connectivity
 		if networkConf.Connectivity == "" {
-			if err := h.setUnknown(nad, networkConf); err != nil {
-				return nil, err
+			if err := h.initializeConnectivity(nad, networkConf); err != nil {
+				klog.Errorf("initialize connectivity of nad %s/%s failed, error: %v", nad.Namespace, nad.Name, err)
```
Comment (gitlawr) [inline pkg/controller/manager/nad/controller.go:116]: Why don't we return the error here?

Comment (yaocw2020) [reply pkg/controller/manager/nad/controller.go:116]: It makes no severely bad effect if initialize connectivity failed. Moreover, the controller will check and set the connectivity every 15 minutes.

**Review by gitlawr — APPROVED** [review]
LGTM

File: pkg/controller/manager/nad/controller.go [hunk 2]
```diff
@@ -353,6 +366,8 @@ func (c *checkMap) addItem(namespace, name, addr string) {
 		c.items[nn] = addr
 		c.mutex.Unlock()
 	}
+
+	klog.Infof("add namespace/name: %+v, itemMap: %+v", nn, c.items)
```
Comment (guangbochen) [inline pkg/controller/manager/nad/controller.go:370]: prefer to log it as a bug log.

File: pkg/controller/manager/nad/controller.go [hunk 3]
```diff
@@ -366,4 +381,6 @@ func (c *checkMap) removeItem(namespace, name string) {
 		delete(c.items, nn)
 		c.mutex.Unlock()
 	}
+
+	klog.Infof("remove namespace/name: %+v, itemMap: %+v", nn, c.items)
```
Comment (guangbochen) [inline pkg/controller/manager/nad/controller.go:385]: ditto

File: pkg/controller/manager/nad/controller.go [hunk 4]
```diff
@@ -288,21 +271,65 @@ func (h Handler) CheckConnectivity() {
...
 	pinger.Count = defaultPingTimes
+	pinger.Timeout = defaultPingTimeout
 	if err := pinger.Run(); err != nil {
-		return connectivity, err
+		return connectivity, fmt.Errorf("ping gw %s failed, error: %w", gw, err)
 	} // blocks until finished
 	stats := pinger.Statistics()
 
-	if stats.PacketsSent != stats.PacketsRecv {
+	klog.Infof("stats: %+v", stats)
```
Comment (guangbochen) [inline pkg/controller/manager/nad/controller.go:330]: prefer to log it as a debug message.

**Review by guangbochen — APPROVED** [review]
LGTM

**Pull request merged**